mlflow/utils/rest_utils.py

```diff
@@ -14,26 +14,76 @@
 
-_DEFAULT_HEADERS = {"User-Agent": "mlflow-python-client/%s" % __version__}
+def _get_http_response_with_retries(max_retries, backoff_factor, retry_codes, **kwargs):
+    """
+    Performs an HTTP request using Python's `requests` module with an automatic retry policy.
+
+    :max_retries: Maximum total number of retries.
+    :backoff_factor: a time factor for exponential backoff. e.g. value 5 means the HTTP request
+      will be retried with interval 5, 10, 20... seconds. A value of 0 turns off the exp-backoff.
+    :retry_codes: a list of HTTP response error codes that qualifies for retry.
+    :kwargs: Keyword arguments to pass to `requests.Session.request()`
+    """
+    assert 0 <= max_retries < 10
+    assert 0 <= backoff_factor < 120
+    retry = Retry(
+        total=max_retries,
+        connect=max_retries,
+        read=max_retries,
+        redirect=max_retries,
+        status=max_retries,
+        status_forcelist=retry_codes,
+        backoff_factor=backoff_factor,
+    )
+    adapter = HTTPAdapter(max_retries=retry)
+    with requests.Session() as http:
+        http.mount("https://", adapter)
+        http.mount("http://", adapter)
+        method = kwargs.pop("method")
+        url = kwargs.pop("url")
+        response = http.request(method, url, **kwargs)
+        return response
 
 
 def http_request(
-    host_creds, endpoint, retries=3, retry_interval=3, max_rate_limit_interval=60, **kwargs
+    host_creds,
+    endpoint,
+    max_retries=3,
+    backoff_factor=5,
+    retry_codes=(429, 500, 503),
+    timeout=10,
+    **kwargs
 ):
     """
     Makes an HTTP request with the specified method to the specified hostname/endpoint. Ratelimit
-    error code (429) will be retried with an exponential back off (1, 2, 4, ... seconds) for at most
-    `max_rate_limit_interval` seconds.  Internal errors (500s) will be retried up to `retries` times
-    , waiting `retry_interval` seconds between successive retries. Parses the API response
+    error code (429), service unavailable code (503) and internal errors (500) are retried with an
+    exponential back off with backoff_factor * (1, 2, 4, ... seconds). Parses the API response
     (assumed to be JSON) into a Python object and returns it.
 
-    :param host_creds: A :py:class:`mlflow.rest_utils.MlflowHostCreds` object containing
+    :host_creds: A :py:class:`mlflow.rest_utils.MlflowHostCreds` object containing
         hostname and optional authentication.
+    :endpoint: a string for service endpoint, e.g. "/path/to/object".
+    :max_retries: maximum number of retries before throwing an exception.
+    :backoff_factor: a time factor for exponential backoff. e.g. value 5 means the HTTP request
+      will be retried with interval 5, 10, 20... seconds. A value of 0 turns off the exp-backoff.
+    :retry_codes: a list of HTTP response error codes that qualifies for retry.
+    :timeout: wait for timeout seconds for response from remote server for connect and read request.
```

Comment:
Added timeout = 10s to allow quicker failure on connection / read timeout. Otherwise it's indefinite by default.

Reply:
Awesome! Can we add this to `cloud_storage_http_request` too?

Reply:
Done.